File: NextcloudTalk/Maps/GeoLocationRichObject.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "GeoLocationRichObject.h"

NSString * const GeoLocationRichObjectType = @"geo-location";

@implementation GeoLocationRichObject

+ (instancetype)geoLocationRichObjectWithLatitude:(double)latitude longitude:(double)longitude name:(NSString *)name
{
    GeoLocationRichObject *object = [[self alloc] init];
    NSString *latitudeString = [[NSNumber numberWithDouble:latitude] stringValue];
    NSString *longitudeString = [[NSNumber numberWithDouble:longitude] stringValue];
    object.objectType = GeoLocationRichObjectType;
    object.objectId = [NSString stringWithFormat:@"geo:%@,%@", latitudeString, longitudeString];
    object.latitude = latitudeString;
    object.longitude = longitudeString;
    object.name = name;
    return object;
}

+ (instancetype)geoLocationRichObjectFromMessageLocationParameter:(NCMessageLocationParameter *)parameter
{
    GeoLocationRichObject *richObject = [[self alloc] init];
    richObject.objectType = parameter.type;
    richObject.objectId = parameter.parameterId;

    if ([parameter.latitude isKindOfClass:[NSNumber class]]) {
        richObject.latitude = [(NSNumber *)parameter.latitude stringValue];
    } else {
        richObject.latitude = parameter.latitude;
    }

    if ([parameter.longitude isKindOfClass:[NSNumber class]]) {
        richObject.longitude = [(NSNumber *)parameter.longitude stringValue];
    } else {
        richObject.longitude = parameter.longitude;
    }

    richObject.name = parameter.name;
    return richObject;
}

- (NSDictionary *)metaData
{
    return @{
             @"latitude": self.latitude,
             @"longitude": self.longitude,
             @"name": self.name
             };
}

- (NSDictionary *)richObjectDictionary
{
    NSError *error;
    NSString *jsonString = @"";
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:[self metaData]
                                                       options:0
                                                         error:&error];
    
    if (!jsonData) {
        NSLog(@"Got an error: %@", error);
    } else {
        jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    }
    
    return @{
             @"objectType": self.objectType,
             @"objectId": self.objectId,
             @"metaData": jsonString
             };
}

@end

File: NextcloudTalk/Maps/GeoLocationRichObject.swift
//
// SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

let geoLocationRichObjectType = "geo-location"

@objcMembers public class GeoLocationRichObject: NSObject {

    public var objectType: String
    public var objectId: String
    public var latitude: String
    public var longitude: String
    public var name: String

    public init(latitude: Double, longitude: Double, name: String) {
        // NSNumber.stringValue renders a whole number as "48", String(Double) would render "48.0" into the objectId
        let latitudeString = NSNumber(value: latitude).stringValue
        let longitudeString = NSNumber(value: longitude).stringValue

        self.objectType = geoLocationRichObjectType
        self.objectId = "geo:\(latitudeString),\(longitudeString)"
        self.latitude = latitudeString
        self.longitude = longitudeString
        self.name = name

        super.init()
    }

    public init(from parameter: NCMessageLocationParameter) {
        self.objectType = parameter.type
        self.objectId = parameter.parameterId
        self.latitude = parameter.latitude ?? ""
        self.longitude = parameter.longitude ?? ""
        self.name = parameter.name

        super.init()
    }

    private var metaData: [String: String] {
        return ["latitude": self.latitude,
                "longitude": self.longitude,
                "name": self.name]
    }

    public func richObjectDictionary() -> [AnyHashable: Any] {
        var jsonString = ""

        do {
            let jsonData = try JSONSerialization.data(withJSONObject: self.metaData)
            jsonString = String(data: jsonData, encoding: .utf8) ?? ""
        } catch {
            NSLog("Got an error: %@", error.localizedDescription)
        }

        return ["objectType": self.objectType,
                "objectId": self.objectId,
                "metaData": jsonString]
    }
}
